kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,270 @@
+local constants = require "kong.constants"
+local clustering_utils = require "kong.clustering.utils"
+-- currently they are the same. But it's possible for we to drop support for old version of DP but keep support of CP
+local supported_services = require "kong.clustering.services.supported"
+local asked_services = require "kong.clustering.services.supported"
+
+local ngx_log = ngx.log
+local ERR = ngx.ERR
+local NOTICE = ngx.NOTICE
+local _log_prefix = "[wrpc-clustering] "
+local table_clear = require "table.clear"
+local table_concat = table.concat
+local lower = string.lower
+
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+local KONG_VERSION
+
+-- it's so annoying that protobuf does not support map to array
+local function wrap_services(service)
+  local wrapped = {}
+  for name, version in pairs(service) do
+    wrapped[name] = {
+      service = version
+    }
+  end
+
+  return wrapped
+end
+
+local function unwrap_services(service)
+  local unwrapped = {}
+  for name, version in pairs(service) do
+    unwrapped[name] = version.service
+  end
+
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for _, ind in ipairs(field) do
+    if type(v) ~= "table" then
```

Comment:
localize `type` `error`

Reply:
pet peeve: `type` is compiled away.  `error` is orders of magnitude more costly than the `_G` access (even more with the string/table/concat operations done here).  no real gain localizing everything in sight.